content/blog/2017/07/2017-07-06-continuousdelivery-devops-artifactory.adoc

```diff
@@ -0,0 +1,287 @@
+
+.pipeline.grovvy (excerpt): Promote Docker image
+[[app-listing]]
+----
+    stage('Promote Docker Image') {
+          sh '''curl -H "X-JFrog-Art-Api:$artifactory_key" -X POST https://$ARTI3/api/docker/docker-local/v2/promote ''' + // <1>
+             '''-H "Content-Type:application/json" ''' + // <2>
+             '''-d \'{"targetRepo" : "docker-prod-local", "dockerRepository" : "michaelhuettermann/tomcat7", "tag": "\'$version\'", "copy": true }\' // <3>
+             '''
+    }
+----
+<1> The shell script to perform the staging of Docker image is based on JFrog REST API.
+<2> Part of parameters are sent in JSON format.
+<3> The payload tells the REST API endpoint what to to, i.e. gives information about _target repo_ and _tag_.
+
+Once the binaries are promoted (and hopefully deployed and tested on respective environments before), we can promote them to become final releases, which I like to call GA.
+
+=== General Availability (GA)
+
+In our scenario, JFrog Bintray serves as the distribution platform to manage and provide binaries for further usage. Bintray can also serve as a Docker registry, or can just
+provide binaries for scripted or manual download. There are again different ways how to promote binaries, in this case from the RC repos inside JFrog Artifactory to the GA storage in JFrog Bintray, and I summarize one of those possible ways. First, let's look at the Jenkins pipeline, showed in the next illustration. The processing is on its way, currently, and we again have a list of linked stages.
+
+image:/images/post-images/artifactory-jenkins/06.png[role=center, title="Promotion to GA is running ..."]
+
+Zooming in now to the key stages, we see that promoting the WAR file is a set of steps that utilize JFrog REST API. We download the binary from JFrog Artifactory, parameterized,
+and upload it to JFrog Bintray.
+
+.pipeline.groovy (excerpt): Promote WAR to Bintray
+[[app-listing]]
+----
+    stage('Promote WAR to Bintray') {
+       steps {
+          sh '''
+             curl -u michaelhuettermann:${bintray_key} -X DELETE https://api.bintray.com/packages/huettermann/meow/cat/versions/$version // <1>
+             curl -u michaelhuettermann:${bintray_key} -H "Content-Type: application/json" -X POST https://api.bintray.com/packages/huettermann/meow/cat/$version --data """{ "name": "$version", "desc": "desc" }""" // <2>
+             curl -T "$WORKSPACE/all-$version-GA.war" -u michaelhuettermann:${bintray_key} -H "X-Bintray-Package:cat" -H "X-Bintray-Version:$version" https://api.bintray.com/content/huettermann/meow/ // <3>
+             curl -u michaelhuettermann:${bintray_key} -H "Content-Type: application/json" -X POST https://api.bintray.com/content/huettermann/meow/cat/$version/publish --data '{ "discard": "false" }' // <4>
+          '''
+       }
+    }
+----
+<1> For testing and demo purposes, we remove the existing release version.
+<2> Next we create the version, in our case it is 1.0.0.
```

Comment:
"in our case", the script never sets `version`, so it could be anything. Perhaps add some fixed values to variables you're using, like `bintray_key`?